### PR TITLE
Driver

### PR DESCRIPTION
**Pull request by daikk115**
Init compute driver for OpenStack and upgrade network driver

Init OPS compute driver (Init base driver follow proposal):

- create
- show
- list
- delete
- shutdown
- start
- reboot
- resize
- add_nic
- delete_nic
- list_nic

Upgrade network driver
- allocate_public_ip

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,4 +1,4 @@
-.. image:: https://travis-ci.org/cloudcomputinghust/CAL.svg?branch=master
+.. image:: https://travis-ci.org/cloudcomputinghust/CAL.svg?branch=driver
```
Comment (hieulq) [inline README.rst]: revert to master branch if you want to merge

File: calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
+""" OpenstackDriver for Compute
+    based on BaseDriver fror Compute Resource
```
Comment (hieulq) [inline calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py]: typo

File: calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+"""
+
+
+import mock
+from keystoneauth1.exceptions.base import ClientException
```
Comment (hieulq) [inline calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py]: pep8 ordering, all external lib import must be separated into another code block, while internal import in another code block.

File: calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    def test_list_nic_unable_to_list(self):
+        self.mock_object(
+            self.fake_driver.client.servers, 'interface_list',
+            mock.Mock(side_effect=ClientException))
+
+        self.assertRaises(ClientException,
+            self.fake_driver.list_nic, 'fake_id')
+
+        self.fake_driver.client.servers.interface_list. \
+            assert_called_once_with('fake_id')
```
Comment (hieulq) [inline calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py]: +1, but in fact I myself do not like this style of mock.

File: calplus/tests/unit/v1/compute/test_client.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
+import mock
+from keystoneauth1.exceptions.base import ClientException
```
Comment (hieulq) [inline calplus/tests/unit/v1/compute/test_client.py]: ditto

File: calplus/v1/compute/client.py [hunk 6]
```diff
@@ -12,3 +12,85 @@ class Client(BaseClient):
     def __init__(self, provider, cloud_config, *args, **kwargs):
         BaseClient.__init__(self, CONF.compute.driver_path,
                             provider, cloud_config)
+
+    def create(self, image, flavor,
+               network_id, name, number, **kargs):
```
Comment (hieulq) [inline calplus/v1/compute/client.py]: this should be kwargs, right?

File: calplus/v1/compute/drivers/base.py [hunk 7]
```diff
@@ -1,9 +1,135 @@
...
+        pass
+
+    @abc.abstractmethod
+    def delete_private_ip(self, instance_id, new_sg):
+        """Delete private IP"""
+        pass
+
+    @abc.abstractmethod
+    def add_public_ip(self, instance_id, new_sg):
+        """Add a external IP"""
```
Comment (hieulq) [inline calplus/v1/compute/drivers/base.py]: an

File: calplus/v1/compute/drivers/base.py [hunk 8]
```diff
@@ -1,9 +1,135 @@
...
+    """docstring for QuotaNetwork"""
+
+    def __init__(self):
+        super(BaseQuota, self).__init__()
+
+    def get(self):
+        """Get quota from Cloud Provider."""
+
+        # get all network quota from Cloud Provider.
+        attrs = ("vcpus",
```
Comment (hieulq) [inline calplus/v1/compute/drivers/base.py]: Does the quota from cloud provider only contain 3 attr described here? Did you miss something?

File: calplus/v1/compute/drivers/openstack.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+    def _setup(self):
+        auth = v3.Password(auth_url=self.auth_url,
+                           user_domain_name=self.user_domain_name,
+                           username=self.username,
+                           password=self.password,
+                           project_domain_name=self.project_domain_name,
+                           project_name=self.project_name)
+        sess = session.Session(auth=auth)
+        self.client = Client("2.1", session=sess)
```
Comment (hieulq) [inline calplus/v1/compute/drivers/openstack.py]: DO NOT hard code the version here, put it in the config

File: calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py (hunk 2 above)
Comment (daikk115) [inline calplus/tests/unit/v1/compute/drivers/test_openstack_driver.py]: https://github.com/cloudcomputinghust/CAL/commit/1803c6cebf43e2fb38fd5ed5a13c87d28bd269e5#diff-e4ec7c2336a8631cc0f5c4f145293979R2

I just fixed it by this commit.

File: README.rst (hunk 1 above)
Comment (daikk115) [inline README.rst]: https://github.com/cloudcomputinghust/CAL/commit/c1883487c5cc05980b534d4e0cf3416890719f7b

**Comment by daikk115** [comment]
@hieulq : Others code review have been fixed by this: https://github.com/cloudcomputinghust/CAL/commit/51f0bc5de949b6688255cdb6d8338640120a8fe8

File: calplus/v1/compute/drivers/base.py (hunk 8 above)
Comment (hieulq) [inline calplus/v1/compute/drivers/base.py]: Please follow our past work in network quota, you need to do a survey.
[https://github.com/cloudcomputinghust/CAL/blob/master/docs/network_nhathqh/abstract_quota.md](url)

File: calplus/v1/compute/drivers/openstack.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        self.user_domain_name = \
+            cloud_config.get('os_project_domain_name', 'default')
+        self.project_domain_name = \
+            cloud_config.get('os_user_domain_name', 'default')
+        self.driver_name = \
+            cloud_config.get('driver_name', 'default')
+        self.tenant_id = cloud_config.get('tenant_id', None)
+        self.limit = cloud_config.get('limit', None)
+        self.client_version = \
+            cloud_config.get('os_novaclient_version', '2.1')
```
Comment (hieulq) [inline calplus/v1/compute/drivers/openstack.py]: @daikk115 Đại THE FALLEN, where the hell did you define all these config options?

Comment (daikk115) [inline calplus/v1/compute/drivers/openstack.py]: humm... we don't have it.

**Comment by hieulq** [comment]
@daikk115 ping?

**Comment by daikk115** [comment]
@hieulq : Coverage decreased by we don't have test for calplus/conf/opts.py
I wonder how to skip coverage on calplus/conf/opts.py or write a testcase for it.
I have try to find something like that but I still didn't get it.

**Pull request closed without merging**